### ISSUE TITLE
[DependencyInjection] Tagged services with negative priority

### ISSUE THREAD

**Issue by Phenix789** [symfony/symfony#50900]
### Symfony version(s) affected

6.3

### Description

By upgrading from 6.2 to 6.3, services implemente an `registerForAutoconfiguration()` interface with negative priority are not sorted correctly.
This seems happen only when `$needsIndexes` is set to true (see below).

### How to reproduce

1. Define a tagged service with a negative priority
```php
$services
  ->set(MyService::class)->tag('controller.argument_value_resolver', ['priority' => -70]);
``` 
2. `ValueResolverInterface` is [`registerForAutoconfiguration`](https://github.com/symfony/symfony/blob/6c85287f18120b6a00014eec806a6427bbb6e080/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php#L603)
3. Now [`ControllerArgumentValueResolverPass`](https://github.com/symfony/symfony/blob/6c85287f18120b6a00014eec806a6427bbb6e080/src/Symfony/Component/HttpKernel/DependencyInjection/ControllerArgumentValueResolverPass.php#L44) call `findAndSortTaggedServices()` with a `TaggedIteratorArgument` and `$needsIndexes` to `true`  instead of a simple `string`.
The behavior of this method is changed and instead of register service only once by [`continue 2`](https://github.com/symfony/symfony/blob/6.4/src/Symfony/Component/DependencyInjection/Compiler/PriorityTaggedServiceTrait.php#L80), the service is registered twice one with -70 priority and one with 0 priority.
5. `uasort()` sort the service with priority 0 before the -70 one (normal)
6. The `$refs` construction add the priority 0 first and the -70 only rewrite it.

### Possible Solution

Locally, I need to register my service with `autoconfigure(false)`.
```php
$services
  ->set(MyService::class)->autoconfigure(false)->tag('controller.argument_value_resolver', ['priority' => -70]);
```

**Comment by xabbuh** [comment]
Can you create a small example application that allows to reproduce your issue?

**Comment by Phenix789** [comment]
I will try to do it tomorrow with simple test.
Thanks

**Comment by Phenix789** [comment]
@xabbuh Hi, you can find a project with the bug here : https://github.com/Phenix789/symfony-project/pull/1
I did a pull request to comment some piece of code if needed.
Thanks

**Comment by Phenix789** [comment]
I also see something.  
If you try to inject services via `#[TaggedLocator]`, services are sorted correctly.
```php
public function __invoke(
  #[TaggedIterator('controller.argument_value_resolver')
  iterable $resolvers,
) {
  dump(iterator_to_array($resolvers));
  exit();
}
```
The bug seems really be in how [`ControllerArgumentValueResolverPass`](https://github.com/symfony/symfony/blob/6c85287f18120b6a00014eec806a6427bbb6e080/src/Symfony/Component/HttpKernel/DependencyInjection/ControllerArgumentValueResolverPass.php#L44) sort tagged services.

**Comment by MatTheCat** [comment]
May be a duplicate of #48019. In that case it would be a documentation issue because duplicated tags caused by autoconfiguration were deemed a feature. As you figured out, disabling autoconfiguration would then be the solution.

**Comment by Phenix789** [comment]
The trouble with this, it's it works in 6.2 and not anymore in 6.3.
In addition, If you ask tagged services via `#[TaggedIterator]` is work fine.
Like say in issue description, bug only appear when `$needsIndexes` is set to true and priority is negative.
I think it's a break change, at least for this tag (`controller.argument_value_resolver`) and without documentation/changelogs.

And disable autoconfiguration for a service can be a bit "painful" when it implement several autoconfigured interfaces (like `LoggerAwareInterface` and others aware interfaces)

**Comment by MatTheCat** [comment]
The change was [the introduction of the `ValueResolver` attribute](https://symfony.com/blog/new-in-symfony-6-3-targeted-value-resolvers), which required to enable `needsIndexes`. Nobody thought it would be a BC break..!

@nicolas-grekas do you know why #35957 only ignores extra autoconfigured tags when `needsIndexes` is `false`?

**Comment by Phenix789** [comment]
I allow myself to up this issue.  
Have you any news ?  
Thanks in advance.

**Comment by carsonbot** [comment]
Hey, thanks for your report!
There has not been a lot of activity here for a while. Is this bug still relevant? Have you managed to find a workaround?

**Comment by carsonbot** [comment]
Hello? This issue is about to be closed if nobody replies.

**Comment by carsonbot** [comment]
Hey,

I didn't hear anything so I'm going to close it. Feel free to comment if this is still relevant, I can always reopen!
